contracts/src/utils/structs/checkpoints.rs

```diff
@@ -0,0 +1,462 @@
+//! Contract module for checkpointing values as they
+//! change at different points in time.
+//! Lets to look up past values by
+//! block number.
+//! See {Votes} as an example.
+//! To create a history of checkpoints,
+//! define a variable type [`Trace160`] in your contract, and store a
+//! new checkpoint for the current transaction block using the
+//! [`Trace160::push`] function.
+use alloy_primitives::{uint, Uint, U256, U32};
+use alloy_sol_types::sol;
+use stylus_proc::{sol_storage, SolidityError};
+use stylus_sdk::storage::{StorageGuard, StorageGuardMut};
+
+use crate::utils::math::alloy::Math;
+
+// TODO#q: add generics for other pairs (uint32, uint224) and (uint48, uint208).
+// Logic should be same.
+type U96 = Uint<96, 2>;
```

Comment:
Currently implemented for key: U96, value: U160 types. Other types has the same logic for implementation and make sense not to copy paste code like it is in solidity. But use generic typing that I'll explore later a bit.